[3.x] Deprecate `NOTIFICATION_MOVED_IN_PARENT`

lawnjelly opened the pull request:
* NOTIFICATION_MOVED_IN_PARENT makes node children management very inefficient.
* Replaced by a NOTIFICATION_CHILD_ORDER_CHANGED (and children_changed signal).
* Most of the previous tasks carried out by NOTIFICATION_MOVED_IN_PARENT are now done not more than a single time per frame.

This PR strictly speaking breaks compatibility, however this notification was very rarely used, even within the engine, and provides an alternate way to do the same. The need to fix this design issue was deemed more important than the likelihood of any breaking in master, and the same applies in 3.x.

Alternative to #74672
Backport of #75701

## Example Project
Run before and after PR:
[ChildOrderChangedBenchmark.zip](https://github.com/godotengine/godot/files/12762158/ChildOrderChangedBenchmark.zip)
_before_ 9069 milliseconds, _after_ 894 milliseconds (approx 10x faster)

The example project stress tests moving sprite children, which causes large numbers of `NOTIFICATION_MOVED_IN_PARENT`. After the PR, these are reduced to once per frame, and the notification is replaced with a cheaper direct function call.

## Explanation
The exponential notification problem remains one of the biggest inefficiencies / achilles heel in the engine in 3.x, so it makes sense to solve it properly, even if there may require some beta testing to iron out any regressions.

This is not a straight backport of #75701 (as 3.x works in a slightly different way) but is inspired to work in a similar(ish) way. There are some significant differences:

* The storage of the dirty parent nodes in `Viewport::GUI` becomes a static rather than _per viewport_. This single list allows storing an ID within the list on the `canvas_item`, which gives O(1) behaviour rather than relying on a `HashSet` and slower lookup as in the 4.x version. It also allows easy flushing once per frame, rather than also flushing on each tick (which is unnecessary and potentially greatly increases the cost).
* As with #74672, the dirty nodes store a minimum and maximum child to update. This makes the processing significantly more efficient when only a small subset of the children are moved. This optimization is not so readily available in master, due to the new way the children are stored.

## Notes
* Like the version in master, this is (strictly speaking) compatibility breaking, because it avoids sending the `NOTIFICATION_MOVED_IN_PARENT` completely. This notification is no longer required in core, but there is the possibility that third parties may rely on it, and need to switch to the new `NOTIFICATION_CHILD_ORDER_CHANGED`. This is however unlikely to be widely used.
* We could still send the old notification if breaking compatibility was a major concern, however this would lose a lot of efficiency. This could for instance be switched on by means of a project setting. However this may be best to avoid unless there are regressions requiring it.  
* Inspired by #75701 this uses simple casting during iteration to determine whether a child needs an update, rather than marking nodes with a flag (the approach used in #74672).
* This potentially makes the `CanvasLayer` updates more efficient because it ensures only a single canvas layer order update per frame.

## Follow up
After we decide whether an approach such as this or the alternative PR is desired, there is another significant optimization which can be made in a followup PR, and possibly also applied in master. Much of the remaining inefficiency is in multiple calls to `VisualServer::canvas_item_set_draw_index()`, where in some cases hundreds or thousands of calls may take place. It seems to make sense to replace this by a single call to an extended version of the function which can change the draw indices of all the items in a single batch.

## Discussion
Personally overall I would be tempted to go with this approach over #74672 . This is now closer to master, and is more efficient because it calls `update_draw_order()` directly instead of going through the heavyweight notifications.
Although it technically breaks compatibility, this is likely to be a rare case, which can either be worked around with the new notification, or we can add an optional legacy mode to call the old notification.
Additionally it should be significantly faster than the version in master because it only updates the range of children changed, rather than _all_ the children every time.

## Update
Due to concerns over applications using custom `MainLoop`, I moved the storage of the dirty list from the `SceneTree` to statics within `Viewport`. This also makes more sense architecturally as the 2D is more localized with other GUI data, and `SceneTree` is not associated with any particular node type.

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

lawnjelly commented on scene/2d/skeleton_2d.cpp:
```diff
@@ -30,9 +30,20 @@
+
+		if (parent) {
+			parent->connect("child_order_changed", this, "_order_changed_in_parent");
+		}
```
This could potentially directly connect the skeleton as in master with a ref counted connection, however doing it via the `Bone2D` as before the PR is less error prone.

Could possibly do a direct skeleton connection in follow up PR.

lawnjelly commented:
Have assigned Juan or Pedro for review, am struggling here because there's not a lot of us who are familiar with this part of core (and less so in 3.x). But anyone else is welcome to review.

Mickeon commented on doc/classes/Node.xml:
```diff
@@ -828,7 +833,7 @@
 		</constant>
 		<constant name="NOTIFICATION_MOVED_IN_PARENT" value="12">
-			Notification received when the node is moved in the parent.
+			This notification is deprecated and is no longer emitted. Use [constant NOTIFICATION_CHILD_ORDER_CHANGED] instead.
```
Matching master
```suggestion
			[i]Deprecated.[/i] This notification is no longer emitted. Use [constant NOTIFICATION_CHILD_ORDER_CHANGED] instead.
```

Mickeon reviewed (commented):
Documentation matches master. It would be nice not to have this PR be forgotten to time, as I feel like 3.6 users would benefit from this a lot, too.

lawnjelly commented on scene/main/scene_tree.h:
```diff
@@ -437,6 +447,9 @@ class SceneTree : public MainLoop {
 	void client_physics_interpolation_remove_spatial(SelfList<Spatial> *p_elem);
 
+	void notify_canvas_parent_children_moved(Node &p_parent, uint32_t p_first_child, uint32_t p_last_child_plus_one);
+	void notify_canvas_parent_child_count_reduced(Node &p_parent);
```
Can you elaborate? I'm not sure I understand reasoning, maybe I'm missing something (then again it's a while since I wrote this PR :grin: ).

AThousandShips replied:
You're passing `Node &` rather than `Node *`, haven't seen that anywhere else

lawnjelly replied:
https://stackoverflow.com/questions/7058339/when-should-i-use-pointers-instead-of-references-in-api-design

https://isocpp.org/wiki/faq/references#:~:text=Use%20references%20when%20you%20can,and%20pointers%20on%20the%20inside.

https://www.geeksforgeeks.org/pointers-vs-references-cpp/

> Use references: 
> In function parameters and return types.
> Use pointers: 
> * If pointer arithmetic or passing a NULL pointer is needed. For example, for arrays (Note that accessing an array is implemented using pointer arithmetic).
> * To implement data structures like a linked list, a tree, etc. and their algorithms. This is so because, in order to point to different cells, we have to use the concept of pointers.

It is usually regarded as good practice in c++ to use references in preference to pointers _wherever possible_. A reference is treated by the compiler as essentially a pointer that CANNOT be NULL. It may also lead to more optimization opportunities in some situations.

This both removes the need for `ERR_FAIL` checks within the function, and prevents the program compiling in a lot of situations that are likely to cause a NULL to be passed, which excludes an entire class of bugs (as you have to either pass an object, or explicitly dereference a pointer on calling, and `*this` is guaranteed to be valid on non-static function as used here).

The reason (as I understand it) Godot uses pointers more often than you would see in most codebases is largely due to _binding_ (for other languages), and the macros which often support pointers but not non-const references. In internal functions that are not bound, reference is preferable imo.

AThousandShips replied:
Oh I'm aware of the benefits 🙂 but it's unorthodox and we do treat our pointers *somewhat* differently than plain ones, but should be safe here

lawnjelly replied:
On the other hand, what we do do normally do is prefix with `r_` for the non-const references.

I have a feeling I didn't change the prefix here because it was just to resolve a compile error with something internally needing to be non-const, I'll double check this.

lawnjelly replied:
`notify_canvas_parent_children_moved()` does need non-const reference, due to calling `set_canvas_parent_id()`. I've changed parameter name to `r_parent` for consistency.

`notify_canvas_parent_child_count_reduced()` turned out to only need a const reference so I've fixed this.

AThousandShips commented on doc/classes/Node.xml:
```diff
@@ -867,6 +872,9 @@
 			Notification received when the node's [NodePath] changed.
 		</constant>
+		<constant name="NOTIFICATION_CHILD_ORDER_CHANGED" value="24">
+			Notification received when the list of children is changed. This happens when child nodes are added, moved or removed.
```
```suggestion
			Notification received when the list of children is changed. This happens when child nodes are added, moved, or removed.
```
Please use [Oxford comma](https://docs.godotengine.org/en/latest/contributing/documentation/docs_writing_guidelines.html#use-the-oxford-comma-to-enumerate-anything).

AThousandShips commented on doc/classes/Node.xml:
```diff
@@ -790,6 +790,11 @@
 		</signal>
+		<signal name="child_order_changed">
+			<description>
+				Emitted when the list of children is changed. This happens when child nodes are added, moved or removed.
```
```suggestion
				Emitted when the list of children is changed. This happens when child nodes are added, moved, or removed.
```

akien-mga reviewed (approved):
Let's give it a spin!

lawnjelly committed:
Deprecate NOTIFICATION_MOVED_IN_PARENT

* NOTIFICATION_MOVED_IN_PARENT makes node children management very inefficient.
* Replaced by a NOTIFICATION_CHILD_ORDER_CHANGED (and children_changed signal).
* Most of the previous tasks carried out by NOTIFICATION_MOVED_IN_PARENT are now done not more than a single time per frame.

This PR breaks compatibility (although this notification was very rarely used, even within the engine), but provides an alternate way to do the same.

lawnjelly commented:
I did a final pass last night / this morning (as I had a rebase disaster and had to look over it minutely anyway :grin: ), and spent a while doing as much testing as possible. I spotted a couple of minor bugs which are now fixed.

I confirmed this gives a *major* speed up in some node operations, but it is a significant change so will be watching for regressions particularly in controls. Hopefully the next beta will give this some decent testing in the wild.
